src/AutoRest.CSharp/LowLevel/AutoRest/DpgOutputLibrary.cs

```diff
@@ -3,36 +3,40 @@
+using AutoRest.CSharp.Common.Input;
 using AutoRest.CSharp.Generation.Types;
 using AutoRest.CSharp.Input;
 
 namespace AutoRest.CSharp.Output.Models.Types
 {
     internal class DpgOutputLibrary : OutputLibrary
     {
+        private readonly IReadOnlyDictionary<InputEnumType, EnumType> _enums;
+        private readonly IReadOnlyDictionary<InputModelType, ModelTypeProvider> _models;
         private readonly TypeFactory _typeFactory;
-        public IReadOnlyList<ModelTypeProvider> Models { get; }
+
+        public IEnumerable<EnumType> Enums => _enums.Values;
+        public IEnumerable<ModelTypeProvider> Models => _models.Values;
         public IReadOnlyList<LowLevelClient> RestClients { get; }
         public ClientOptionsTypeProvider ClientOptions { get; }
 
-        public DpgOutputLibrary(Func<TypeFactory, IReadOnlyList<ModelTypeProvider>> modelsFactory, Func<TypeFactory, IReadOnlyList<LowLevelClient>> restClientsFactory, ClientOptionsTypeProvider clientOptions)
+        public DpgOutputLibrary(Func<TypeFactory, IReadOnlyDictionary<InputEnumType, EnumType>> enumsFactory, Func<TypeFactory, IReadOnlyDictionary<InputModelType, ModelTypeProvider>> modelsFactory, Func<TypeFactory, IReadOnlyList<LowLevelClient>> restClientsFactory, ClientOptionsTypeProvider clientOptions)
         {
             _typeFactory = new TypeFactory(this);
-            Models = modelsFactory(_typeFactory);
+            _enums = enumsFactory(_typeFactory);
+            _models = modelsFactory(_typeFactory);
             RestClients = restClientsFactory(_typeFactory);
             ClientOptions = clientOptions;
         }
 
-        public override CSharpType FindTypeForSchema(Schema schema)
-            => schema.Type switch
-            {
-                AllSchemaTypes.Choice => _typeFactory.CreateType(((ChoiceSchema)schema).ChoiceType, false),
-                AllSchemaTypes.SealedChoice => _typeFactory.CreateType(((SealedChoiceSchema)schema).ChoiceType, false),
-                // This is technically invalid behavior, we are hitting this in generating responses we throw away.
-                // https://github.com/Azure/autorest.csharp/issues/1108
-                // throw new InvalidOperationException($"FindTypeForSchema of invalid schema {schema.Name} in LowLevelOutputLibrary");
-                _ => new CSharpType(typeof(object))
-            };
+        public override CSharpType ResolveEnum(InputEnumType enumType) => _enums != null
+            ? _enums.TryGetValue(enumType with {IsNullable = false}, out var typeProvider) ? typeProvider.Type : _typeFactory.CreateType(enumType.EnumValueType)
+            : throw new InvalidOperationException($"{nameof(ResolveEnum)} is called before enums are generated.");
+        public override CSharpType ResolveModel(InputModelType model) => _models != null
+            ? _models.TryGetValue(model, out var typeProvider) ? typeProvider.Type : new CSharpType(typeof(object), model.IsNullable)
```

Comment:
Maybe use _typeFactory to be consistent?
```suggestion
            ? _models.TryGetValue(model, out var typeProvider) ? typeProvider.Type : _typeFactory.CreateType(model)
```

Reply:
That would create an infinite recursion. This check is for swagger-based DPG that doesn't have models yet. When HLC is switched on the DPG rails, I'll remove this check.